build-support/bin/bootstrap_binary_util.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python2.7
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# Helper for download_binary.sh to use BinaryUtil to download the appropriate binaries.
+#
+# usage: bootstrap_binary_util.py util_name version filename
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import os
+import re
+import sys
+from functools import reduce
+
+from pants.binaries.binary_util import BinaryRequest, BinaryUtil
+from pants.fs.archive import archiver_for_path
+from pants.option.global_options import GlobalOptionsRegistrar
+from pants.option.options_bootstrapper import OptionsBootstrapper
+from pants.subsystem.subsystem import Subsystem
+
+
+def main():
+  util_name, version, filename = tuple(sys.argv[1:])
+  options_bootstrapper = OptionsBootstrapper()
+  subsystems = (GlobalOptionsRegistrar, BinaryUtil.Factory)
+  known_scope_infos = reduce(set.union, (ss.known_scope_infos() for ss in subsystems), set())
+  options = options_bootstrapper.get_full_options(known_scope_infos)
+  for subsystem in subsystems:
+    subsystem.register_options_on_scope(options)
+  Subsystem.set_options(options)
+
+  # If the filename provided ends in a known archive extension (such as ".tar.gz"), then we get the
+  # appropriate Archiver to pass to BinaryUtil.
+  archiver_for_current_binary = None
+  try:
+    archiver_for_current_binary = archiver_for_path(filename)
+    # BinaryRequest requires the `name` field to be provided without an extension, as it appends the
+    # archiver's extension if one is provided, so we have to remove it here.
+    filename = re.sub(
+      r'\.{}\Z'.format(re.escape(archiver_for_current_binary.extension)),
```

Comment:
I had to look up \Z, which was fun, I did not know $ had a synonym or vice-versa - but, if you got an acrhiver back, you know you can just lop off the filename verbatim - perhaps `filename[:-(len(extension)+1)]` or `filename.rsplit('.' + extension, 1)[0]`?

Reply:
Defaults for multiline modes differ across languages and other languages such as emacs lisp have more customized extensions for e.g. symbol boundaries so it's easier for me to be explicit than to wonder whether a stray `$` is going to slurp a whole file in, or worse, make repeated regex applications start at increasingly more incorrect positions. This regex replace was definitely a 3-4 second solution -- either of what you suggested seem more appropriate and I will make that fix.

Reply:
I don’t fully follow this conversation, but if you go with regex consider a very brief comment explaining what Z means

Reply:
Sorry -- "being explicit" refers to using `\Z` compulsively instead of the more recognizable `$` (although if we wish to be truly pedantic, filenames *can* have newlines, so this should really also have a TODO to replace all such filename manipulations with methods in `util` or something so we can contain the pedantry).